examples/step-63/step-63.cc

```diff
@@ -19,16 +19,1091 @@
+    };
+
+    auto copier_mg = [&](const CopyData &copy_data) {
+      boundary_constraints[copy_data.level].distribute_local_to_global(
+        copy_data.cell_matrix,
+        copy_data.local_dof_indices,
+        mg_matrices[copy_data.level]);
+
+      // If (i,j) is an interface_out dof pair, then (j,i) is an interface_in
+      // dof pair. Note: for interface_in, we load the transpose of the
+      // interface entries, i.e., the entry for dof pair (j,i) is stored in
+      // interface_in(i,j).
+      for (unsigned int i = 0; i < copy_data.dofs_per_cell; ++i)
+        for (unsigned int j = 0; j < copy_data.dofs_per_cell; ++j)
+          if (mg_constrained_dofs.is_interface_matrix_entry(
+                copy_data.level,
+                copy_data.local_dof_indices[i],
+                copy_data.local_dof_indices[j]))
+            {
+              mg_interface_out[copy_data.level].add(
+                copy_data.local_dof_indices[i],
+                copy_data.local_dof_indices[j],
+                copy_data.cell_matrix(i, j));
+              mg_interface_in[copy_data.level].add(
+                copy_data.local_dof_indices[i],
+                copy_data.local_dof_indices[j],
+                copy_data.cell_matrix(j, i));
+            }
+    };
+
+    MeshWorker::mesh_loop(dof_handler.begin_mg(),
+                          dof_handler.end_mg(),
+                          cell_worker_mg,
+                          copier_mg,
+                          ScratchData<dim>(fe, quad_degree),
+                          CopyData(),
+                          MeshWorker::assemble_own_cells);
+  }
+
+
+  template <int dim>
+  std::unique_ptr<MGSmoother<Vector<double>>>
+  AdvectionProblem<dim>::create_smoother()
+  {
+    if (settings.smoother_type == "sor")
+      {
+        typedef PreconditionSOR<SparseMatrix<double>> Smoother;
+
+        auto smoother =
+          std_cxx14::make_unique<MGSmootherPrecondition<SparseMatrix<double>,
+                                                        Smoother,
+                                                        Vector<double>>>();
+        smoother->initialize(mg_matrices,
+                             Smoother::AdditionalData(fe.degree == 1 ? 1.0 :
+                                                                       0.62));
+        smoother->set_steps(2);
+        return smoother;
+      }
+    else if (settings.smoother_type == "jacobi")
+      {
+        typedef PreconditionJacobi<SparseMatrix<double>> Smoother;
+        auto                                             smoother =
+          std_cxx14::make_unique<MGSmootherPrecondition<SparseMatrix<double>,
+                                                        Smoother,
+                                                        Vector<double>>>();
+        smoother->initialize(mg_matrices,
+                             Smoother::AdditionalData(fe.degree == 1 ? 0.6667 :
+                                                                       0.47));
```

Comment:
I look forward to the justification for the values here and above :-)

Reply:
yes... These were picked just by trial and error. We were actually curious how we would explain them..

Reply:
The old "DEAL" library that predated deal.II had a few of these constants hard-coded in various places. The commentary in DEAL was generally substantially sparser than in deal.II, but for such "magic" cases, there were occasional jokes attached. 

In general, saying that they were found to be best based on trial and error is legitimate.